src/lib/xmlparse.rs

```diff
@@ -1220,86 +1220,43 @@ pub const EXPAND_SPARE: c_int = 24;
 pub const INIT_SCAFFOLD_ELEMENTS: c_int = 32;
 
 macro_rules! MALLOC {
-    ($parser:expr, $size:expr $(,)?) => {
-        (*$parser)
-            .m_mem
-            .malloc_fcn
-            .expect("non-null function pointer")($size)
-    };
+    ($size:expr $(,)?) => {{
+        let layout = Layout::from_size_align($size as usize, 1)
+            .expect("failed to create Layout");
+        alloc::alloc(layout) as *mut c_void
+    }};
+    // FIXME: we need the @ to disambiguate from the previous form
+    (@$ty:ty) => {{
+        let layout = Layout::new::<$ty>();
+        alloc::alloc(layout) as *mut $ty
+    }};
+    [$ty:ty; $n:expr] => {{
+        let layout = Layout::array::<$ty>($n as usize)
+            .expect("failed to create array Layout");
+        alloc::alloc(layout) as *mut $ty
+    }};
 }
 macro_rules! REALLOC {
-    ($parser:path, $ptr:expr, $size:expr $(,)?) => {
-        (*$parser)
-            .m_mem
-            .realloc_fcn
-            .expect("non-null function pointer")($ptr, $size)
-    };
+    ($ptr:expr, $size:expr $(,)?) => {{
+        let layout = Layout::from_size_align($size as usize, 1)
+            .expect("failed to create Layout");
+        alloc::realloc($ptr as *mut u8, layout, $size as usize) as *mut c_void
+    }};
+    ($ptr:expr => [$ty:ty; $n:expr]) => {{
+        // FIXME: we should pass the old `Layout` to `realloc`,
+        // but we don't have it
+        let layout = Layout::array::<$ty>($n as usize)
+            .expect("failed to create array Layout");
+        alloc::realloc($ptr as *mut u8, layout, layout.size()) as *mut $ty
+    }};
 }
 macro_rules! FREE {
-    ($parser:expr, $ptr:expr $(,)?) => {
-        (*$parser)
-            .m_mem
-            .free_fcn
-            .expect("non-null function pointer")($ptr)
+    ($ptr:expr $(,)?) => {
+        // FIXME: get the actual layout somehow
+        alloc::dealloc($ptr as *mut u8, Layout::new::<u8>())
```

Comment:
What happens if someone swaps the global allocator on us between allocation and freeing?

Reply:
Bad things happen. I don't have a good way of handling that case, and I don't think Rust does either. The global allocator is supposed to be a singleton.